server/pulp/server/lazy/url.py

```diff
@@ -0,0 +1,564 @@
+    """
+
+    @staticmethod
+    def decode(encoded):
+        """
+        Decode the URL formatted parameters into a dictionary.
+
+        :param encoded: A URL formatted parameters.
+        :type encoded: str
+        :return: The parameters as a dictionary.
+        :rtype: dict
+        """
+        decoded = {}
+        for pair in encoded.split(';'):
+            if not pair:
+                continue
+            k, v = pair.split('=')
+            decoded[k] = unquote(v)
+        return decoded
+
+    @staticmethod
+    def encode(parsed):
+        """
+        Join the parsed parameters dictionary into URL parameter format.
+
+        :param parsed: A parsed parameters.
+        :type parsed: dict
+        :return: A parameters string.  Eg: name=john;age=10
+        :rtype: str
+        """
+        encoded = []
+        for k, v in parsed.items():
+            p = '{k}={v}'.format(k=k, v=quote(v))
+            encoded.append(p)
+        return ';'.join(encoded)
+
+
+class Key(object):
+    """
+    Provides RSA key management.
+    """
+
+    @staticmethod
+    def load(path=None, pem=None):
+        """
+        Get/Load an RSA key at the specified path.
+
+        :param path: An absolute path to a PEM encoded key.
+        :type path: str
+        :param pem: A PEM encoded key.
+        :type pem: str
+        :return: The loaded key.
+        :rtype: RSA.RSA
+        """
+        if path:
+            with open(path) as fp:
+                pem = fp.read()
+        bfr = BIO.MemoryBuffer(pem)
+        if 'PRIVATE' in pem:
+            key = RSA.load_key_bio(bfr)
+        else:
+            key = RSA.load_pub_key_bio(bfr)
+        return key
+
+
+class URL(object):
+    """
+    An URL object that supports signing.
+
+    :ivar content: The actual URL.
+    :type content: ParseResult
+    """
+
+    POLICY = 'policy'
+    SIGNATURE = 'signature'
+
+    def __init__(self, content):
+        """
+        :param content: The *actual* URL key.
+        :type content: str
+        """
+        self.content = urlparse(content)
+
+    @property
+    def scheme(self):
+        """
+        :return: The *scheme* component of the URL.
+        :rtype: str
+        """
+        return self.content.scheme
+
+    @property
+    def netloc(self):
+        """
+        :return: The *host:port* component of the URL.
+        :rtype: str
+        """
+        return self.content.netloc
+
+    @property
+    def path(self):
+        """
+        :return: The *path* component of the URL.
+        :rtype: str
+        """
+        return self.content.path
+
+    @property
+    def params(self):
+        """
+        :return: The *params* component of the URL.
+        :rtype: str
+        """
+        return self.content.params
+
+    @property
+    def query(self):
+        """
+        :return: The *query* component of the URL.
+        :rtype: str
+        """
+        return self.content.query
+
+    @property
+    def uri(self):
```

Comment:
Nitpicky, but I got to wondering what the difference is, exactly, between a URL and a URI. Going off of http://www.ietf.org/rfc/rfc3986.txt section 1.1.3, all URLs are URIs. Looking at the code you're using it to mean the path and any parameters and query which doesn't seem to match up with my understanding of a URI. That being said, I'm not sure what the best thing to call this is.